Jenkinsfile

```diff
@@ -0,0 +1,246 @@
+        }
+        else {
+            env.GIT_COMMIT_HASH = bat(script: 'git rev-parse HEAD', returnStdout: true).trim()
+        }
+    }
+}
+
+
+def copyArtifacts(){
+    echo "Copying artifacts from mbed-client-cliapp_for_CI/label=GCC_ARM"
+    step([
+        $class: 'CopyArtifact',
+        fingerprintArtifacts: true,
+        flatten: true,
+        projectName: "mbed-client-cliapp_for_CI/label=GCC_ARM",
+        filter: '**/mbed-client-cliapp.bin',
+        selector: [
+            $class: 'StatusBuildSelector',
+            stable: true
+        ],
+        target: 'test/'
+    ])
+}
+
+
+def setBuildStatus(String state, String context, String message) {
+    step([
+        $class: "GitHubCommitStatusSetter",
+        reposSource: [
+            $class: "ManuallyEnteredRepositorySource",
+            url: "https://github.com/ARMmbed/mbed-clitest.git"
+        ],
+        contextSource: [
+            $class: "ManuallyEnteredCommitContextSource",
+            context: context
+        ],
+        errorHandlers: [[
+            $class: "ChangingBuildStatusErrorHandler",
+            result: "UNSTABLE"
+        ]],
+        commitShaSource: [
+            $class: "ManuallyEnteredShaSource",
+            sha: env.GIT_COMMIT_HASH
+        ],
+        statusResultSource: [
+            $class: 'ConditionalStatusResultSource',
+            results: [
+                [
+                    $class: 'AnyBuildResult',
+                    message: message,
+                    state: state
+                ]
+            ]
+        ]
+     ])
+}
+
+
+def unittest(String pythonVersion) {
+    echo "set ${pythonVersion} github status"
+    String buildName = "linux ${pythonVersion} test"
+    setBuildStatus('PENDING', "${buildName}", 'test start')
+    try {
+        if (pythonVersion == "py3") {
+            // create python 3 venv
+            sh """
+            python3 -m venv .py3venv --without-pip
+            . .py3venv/bin/activate
+            curl https://bootstrap.pypa.io/get-pip.py | python
+            id
+            python setup.py install
+            pip install coverage mock
+            pip install -I mbed-ls==1.2.14
+            coverage run -m unittest discover -s test -vvv
+            coverage html
+            coverage xml
+            deactivate
+            """
+        } else {
+            // create python2 venv, do installation and run tests
+            sh """
+            virtualenv --python=../usr/bin/python .py2venv --no-site-packages
+            . .py2venv/bin/activate
+            id
+            python setup.py install
+            pip install coverage mock
+            coverage run -m unittest discover -s test -vvv
+            coverage html
+            coverage xml
+            deactivate
+            """
+        }
+        setBuildStatus('SUCCESS', "${buildName}", "test done")
+    } catch (err) {
+        setBuildStatus('FAILURE', "${buildName}", "test failed")
+        currentBuild.result = 'FAILURE'
```

Comment:
I'd suggest just rethrowing the exception after setting the build status. If an abort was issued while the script was running this could potentially result in the build result being set to 'FAILURE' rather than 'ABORTED'. Rethrowing lets Jenkins handle it as usual.

Reply:
throw is not good, as if py2 fails, we still want to execute py3. instead of just stop this build.